lib/internal/http2/compat.js

```diff
@@ -665,24 +659,26 @@ class Http2ServerResponse extends Stream {
       encoding = 'utf8';
     }
 
-    if (chunk !== null && chunk !== undefined)
-      this.write(chunk, encoding);
-
-    const isFinished = this.finished;
-    state.headRequest = stream.headRequest;
-    state.ending = true;
-
     if (typeof cb === 'function') {
-      if (isFinished)
+      if (stream.writableFinished)
         this.once('finish', cb);
       else
         stream.once('finish', cb);
     }
 
+    if (state.ending) {
+      return this;
```

Comment:
This check had to be moved here in order to ensure `cb` is invoked on `finish` even if already ending.